### PR TITLE
Multi-version compatibility including 7.1

### PR DESCRIPTION
**Pull request by quentindemetz**
Fork the directory tree by activerecord version and load the proper definition of the adapter based on the runtime version of ActiveRecord.

This includes the changes in https://github.com/pennylane-hq/activerecord7-redshift-adapter/pull/13

https://github.com/makandra/gemika seems to be a good way to implement basic testing for this

### CODE REVIEW

**Review by tdeo — APPROVED** [review]

File: lib/active_record/connection_adapters/redshift_adapter.rb [hunk 1]
```diff
@@ -1,770 +1,9 @@
...
-      end
-    end
-  end
-end
+if ActiveRecord.version >= Gem::Version.new('7.1.0')
+  require_relative 'redshift_7_1_adapter'
+elsif ActiveRecord.version >= Gem::Version.new('7.0.0')
+  require_relative 'redshift_7_0_adapter'
+else
+  raise ArgumentError
```
Comment (owst) [inline lib/active_record/connection_adapters/redshift_adapter.rb]: I think:

1. This should just be a `RuntimeError` - there's no argument passed to be wrong/invalid
2. We should include a descriptive message to say what's gone wrong

File: lib/active_record/connection_adapters/redshift_7_1_adapter.rb [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+          end
+          @statements[sql_key]
+        end
+      end
+
+      # Connects to a PostgreSQL server and sets up the adapter depending on the
+      # connected server's characteristics.
+      def connect
+        @raw_connection = PG.connect(@raw_connection_parameters)
+        @connection = @raw_connection if ::Rails::VERSION::MINOR == 0
```
Comment (owst) [inline lib/active_record/connection_adapters/redshift_7_1_adapter.rb]: This line can be removed now that the 7.1 adaptor isn't used for 7.0

**Review by owst — CHANGES_REQUESTED** [review]
Interesting idea @quentindemetz! 

This is in contrast to the approach of maintaining separate branches (and building/releasing separate gems) for the different supported versions (possibly just Major.Minor of the corresponding Rails version) - I'm not sure either way about the tradeoffs in terms of test complexity or if there's any overhead to including (largely duplicated) code in the gem

**Comment by quentindemetz** [comment]
Thanks for your feedback @owst 
- I don't see any overhead in including largely duplicated code, since only the relevant version is required at runtime
- Testing is obviously still TBD but 🤞🏻 for https://github.com/makandra/gemika
- I'd rather not have to maintain x branches and backport fixes to all branches, and have everything live in one place. In the end it's the same result (many versions of the code are "live" and usable by users) and I find that having everything on the same branch is likely to be simpler operationally 🤞🏻

**Review by owst — APPROVED** [review]
This looks good to me. It would be nice to try and obtain https://rubygems.org/gems/activerecord-redshift-adapter or a similar version-less gem so that this can become the go-to gem for Rails 7 and beyond

**Comment by quentindemetz** [comment]
Playing devil's advocate, but if we get `activerecord-redshift-adapter` it will break existing (legacy and non-updated) rails 3 apps which might rely on this gem. There's likely a bunch of these out there. 

How about `activerecord-adapter-redshift` 😉 ?

**Comment by owst** [comment]
That's a good point - I wonder if a major version bump would be sufficient to sign post that the gem has significantly changed. Either way, it's moot unless rubygems would allow publishing a new gem under and old name - I somewhat doubt it. 

I think it would be better to keep the `activerecord-X-adapter` naming convention along with the other popular gems:

image

So perhaps `activerecord-redshift-X-adapter` for some reasonable choice of X - new? modern? recent? ...?

**Pull request merged**